go/cmd/dolt/commands/filter-branch.go

```diff
@@ -0,0 +1,241 @@
+func (cmd FilterBranchCmd) Name() string {
+	return "filter-branch"
+}
+
+// Description returns a description of the command
+func (cmd FilterBranchCmd) Description() string {
+	return filterBranchDocs.ShortDesc
+}
+
+// CreateMarkdown creates a markdown file containing the helptext for the command at the given path
+func (cmd FilterBranchCmd) CreateMarkdown(fs filesys.Filesys, path, commandStr string) error {
+	ap := cmd.createArgParser()
+	return CreateMarkdown(fs, path, cli.GetCommandDocumentation(commandStr, filterBranchDocs, ap))
+}
+
+func (cmd FilterBranchCmd) createArgParser() *argparser.ArgParser {
+	ap := argparser.NewArgParser()
+	ap.SupportsFlag(allFlag, "a", "filter all branches")
+	return ap
+}
+
+// EventType returns the type of the event to log
+// todo: make event
+//func (cmd FilterBranchCmd) EventType() eventsapi.ClientEventType {
+//	return eventsapi.ClientEventType_LS
+//}
+
+// Exec executes the command
+func (cmd FilterBranchCmd) Exec(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := cmd.createArgParser()
+	help, usage := cli.HelpAndUsagePrinters(cli.GetCommandDocumentation(commandStr, filterBranchDocs, ap))
+	apr := cli.ParseArgs(ap, args, help)
+
+	if apr.NArg() != 1 {
+		args := strings.Join(apr.Args(), ", ")
+		verr := errhand.BuildDError("%s takes exactly 1 arg, %d provided: %s", cmd.Name(), apr.NArg(), args).Build()
+		return HandleVErrAndExitCode(verr, usage)
+	}
+
+	notFound := make(missingTbls)
+	query := apr.Arg(0)
+	replay := func(ctx context.Context, root, _, _ *doltdb.RootValue) (*doltdb.RootValue, error) {
+		return processFilterQuery(ctx, dEnv, root, query, notFound)
+	}
+
+	var err error
+	if apr.Contains(allFlag) {
+		err = rebase.AllBranches(ctx, dEnv, replay, rebase.EntireHistory)
+	} else {
+		err = rebase.CurrentBranch(ctx, dEnv, replay, rebase.EntireHistory)
+	}
+	if err != nil {
+		return HandleVErrAndExitCode(errhand.VerboseErrorFromError(err), usage)
+	}
+
+	for h, e := range notFound {
+		cli.PrintErrln(color.YellowString("for root value %s: %s", h.String(), e.Error()))
+	}
+
+	return 0
+}
+
+type missingTbls map[hash.Hash]*errors.Error
+
+func processFilterQuery(ctx context.Context, dEnv *env.DoltEnv, root *doltdb.RootValue, query string, mt missingTbls) (*doltdb.RootValue, error) {
+	sqlCtx, se, err := monoSqlEngine(ctx, dEnv, root)
+	if err != nil {
+		return nil, err
+	}
+
+	sqlStatement, err := sqlparser.Parse(query)
+	if err != nil {
+		return nil, err
+	}
+
+	rh, err := root.HashOf()
+	if err != nil {
+		return nil, err
+	}
+
+	itr := sql.RowsToRowIter()
+	switch s := sqlStatement.(type) {
+	case *sqlparser.Delete:
+		ok := se.checkThenDeleteAllRows(sqlCtx, s)
+		if !ok {
+			_, itr, err = se.query(sqlCtx, query)
+		}
+	default:
+		// todo: support insert, update, ddl
+		return nil, fmt.Errorf("SQL statement not supported for filter-branch: '%v'.", query)
+	}
+
+	err, ok := captureTblNotFoundErr(err, mt, rh)
```

Comment:
Ideas for interfaces for this in the future? We can allow the query to parameterize based on the commit we're operating on...we can try to distinguish between logical and internal|i/o errors and have a mode where we continue across any logical errors...we can allow the user to give a class of logical errors which should be converted to warnings...not so easy...

Reply:
I think we should handle the class of logical errors stemming from schema changes and error on everything else. 
Specifically, we should rewrite queries to handle column add/drops and table/column renames so that we always execute something useful when appropriate. 
This version doesn't have an ancestor commit option to stop at, so I thought we should swallow missing table errors. I think that's probably the right thing to do in general.